src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,210 @@
+                {
+                    BoundedCapacity = 1000,
+                    MaxDegreeOfParallelism = Environment.ProcessorCount * 8,
+                    CancellationToken = cancellationToken,
+                });
+
+            var downloadBlock = NonSwallowingActionBlock.Create<(Stream stream, string targetPath)>(
+                item =>
+                {
+                    using (item.stream)
+                    {
+                        int retryCount = 0;
+                        try
+                        {
+                            using (var fileStream = new FileStream(item.targetPath, FileMode.Create))
+                            {
+                                item.stream.CopyTo(fileStream);
+                            }
+                        }
+                        catch (IOException exception) when (retryCount < 3)
+                        {
+                            Console.WriteLine($"Exception caught: {exception.Message}, on retry count {retryCount}, Retrying\n");
+                            retryCount++;
+                        }
+                    }               
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 1000,
+                    MaxDegreeOfParallelism = Environment.ProcessorCount,
+                    CancellationToken = cancellationToken,
+                });
+
+            batchItemsBlock.LinkTo(fetchStream, new DataflowLinkOptions() { PropagateCompletion = true });
+            fetchStream.LinkTo(downloadBlock, new DataflowLinkOptions() { PropagateCompletion = true });
+
+            await batchItemsBlock.SendAllAndCompleteAsync(fileItems, downloadBlock, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<Stream> DownloadFileFromContainerAsync(Tuple<long, string> containerIdAndRoot, Guid scopeIdentifier, FileContainerHttpClient containerClient, FileContainerItem item, string targetPath, CancellationToken cancellationToken)
+        {
+            Stream responseStream = await AsyncHttpRetryHelper.InvokeAsync(
+                async () =>
+                {
+                    Stream internalResponseStream = await containerClient.DownloadFileAsync(containerIdAndRoot.Item1, item.Path, cancellationToken, scopeIdentifier).ConfigureAwait(false);
+                    return internalResponseStream;
+                },
+                maxRetries: 5,
+                cancellationToken: cancellationToken,
+                tracer: this.tracer,
+                continueOnCapturedContext: false
+                );
+
+            return responseStream;
+        }
+
+        private string ResolveTargetPath(string rootPath, FileContainerItem item, string artifactName)
+        {
+            var index = artifactName.Length;
+            var itemPathWithoutDirectoryPrefix = (index != -1 && index < item.Path.Length) ? item.Path.Substring(index + 1) : string.Empty;
```

Comment:
>itemPathWithoutDirectoryPrefix [](start = 16, length = 30)

You can use Path.GetFileName instead of doing this indexing logic:
[https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfilename?view=netframework-4.8](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfilename?view=netframework-4.8)